warehouse/slave.go

```diff
@@ -135,51 +136,69 @@ func (job *PayloadT) getColumnName(columnName string) string {
 	return warehouseutils.ToProviderCase(job.DestinationType, columnName)
 }
 
-// TODO: work in progress
+type loadFileUploadJob struct {
+	tableName  string
+	outputFile misc.GZipWriter
+}
+
 func (jobRun *JobRunT) uploadLoadFilesToObjectStorage() ([]int64, error) {
 	job := jobRun.job
 	uploader, err := job.getFileManager()
 	if err != nil {
 		return []int64{}, err
 	}
 	var loadFileIDs []int64
-	var loadFileIDChan chan int64
-	var loadFileErrChan chan error
-	maxParallelLoadsChan := make(chan struct{}, 3) // TODO: add config variable
-	// Upload each generated load file to ObjectStorage
-	// On successful upload, store the saved fileID in wh_load_files table
-	for tableName, outputFile := range jobRun.outputFileWritersMap {
-		tableName := tableName
-		outputFile := outputFile
-		maxParallelLoadsChan <- struct{}{}
-		rruntime.Go(func() {
-			uploadOutput, err := jobRun.uploadLoadFileToObjectStorage(uploader, &outputFile, tableName)
-			if err != nil {
-				loadFileErrChan <- err
-				return
+	loadFileIDChan := make(chan int64, len(jobRun.outputFileWritersMap))
+	uploadJobChan := make(chan *loadFileUploadJob, 100)
+	uploadErrorChan := make(chan error)
+	ctx, cancel := context.WithCancel(context.Background())
+	for i := 0; i < numLoadFileUploadWorkers; i++ {
+		go func(ctx context.Context) {
+			for uploadJob := range uploadJobChan {
+				select {
+				case <-ctx.Done():
+					pkgLogger.Debugf("context is cancelled, stopped processing load file for staging file id %s ", job.StagingFileID)
```

Comment:
We should use Errorf here?

Reply:
we are logging error when we are receiving error from the err channel